atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,222 @@
+     * the new resource.
+     *
+     * <b>Do not</b> block on close or update the refreshable within the scope of runWith. Doing so will result in a
+     * deadlock.
+     *
+     * There are no timeouts applied to these invocations. If required, consider using the
+     * {@link com.google.common.util.concurrent.TimeLimiter} utilities. Long-running invocations will block close.
+     *
+     * @param task The function to execute using the current resource. This function may throw a checked exception,
+     *             and this exception will propagate.
+     * @param <V> The return type of the task
+     * @param <K> The checked exception type, if any. If multiple checked exceptions are thrown, this will resolve to
+     *          the lowest common ancestor, which may be {@link Exception}
+     * @return The result of `task`
+     * @throws K Any checked exceptions thrown by the underlying task.
+     */
+    public <V, K extends Exception> V runWithResource(FunctionCheckedException<T, V, K> task) throws K {
+        return runIfNotClosed(() -> {
+            Pair<ReadWriteLock, T> resourceWithLock = refreshableResourceWithLock.get();
+            Lock readLock = resourceWithLock.getLhSide().readLock();
+            return runWithLock(readLock, () -> task.apply(resourceWithLock.getRhSide()));
+        });
+    }
+
+    /**
+     * Utility method for {@link #runWithResource} for tasks that have no return value.
+     *
+     * See {@link #runWithResource} for semantics, guarantees and conditions.
+     *
+     */
+    public <K extends Exception> void executeWithResource(ConsumerCheckedException<T, K> task) throws K {
+        runWithResource(resource -> {
+            task.accept(resource);
+            return null;
+        });
+    }
+
+    private static <V, K extends Exception> V runWithLock(Lock lock, CallableCheckedException<V, K> task) throws K {
+        lock.lock();
+        try {
+            return task.call();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private static <K extends Exception> void runWithLock(Lock lock, RunnableCheckedException<K> task) throws K {
+        runWithLock(lock, () -> {
+            task.run();
+            return null;
+        });
+    }
+
+    private <V, K extends Exception> V runIfNotClosed(CallableCheckedException<V, K> task) throws K {
+        Lock readLock = isClosedLock.readLock();
+        return runWithLock(readLock, () -> {
+            if (isClosed) {
+                // This exception bubbles up to the caller of #runWithResource, and blocks an update for the
+                // Refreshable#map call.
+                throw new SafeIllegalStateException("Attempted to use resource after the container was closed.");
+            }
+            return task.call();
+        });
+    }
+
+    @FunctionalInterface
+    public interface ConsumerCheckedException<T, K extends Exception> {
+        void accept(T input) throws K;
```

Comment:
This isn't in `atlasdb-commons` for the same reason `CallableCheckedException` wasn't originally - there's only one class using this right now, so there was little reason to share this, at least as of now.